dbEvent: correct eventsRemaining

mdavidsaver opened the pull request:
Only pass eventsRemaining!=0 if no pending events have been canceled.  Attempts a partial fix #324 without excess message fragmentation.

mdavidsaver commented on modules/database/src/ioc/db/dbEvent.c:
```diff
@@ -1009,7 +1011,7 @@ static int event_read ( struct event_que *ev_que )
                 /* Issue user callback */
                 ( *user_sub ) ( pevent->user_arg, pevent->chan,
-                                ev_que->evque[ev_que->getix] != EVENTQEMPTY, pfl );
+                                eventsRemaining, pfl );
```
Looking again at the previous code.  `ev_que->evque[ev_que->getix]` is being evaluated without locking, so a race condition as well.

ralphlange commented:
Thanks!
I will prepare a test library with this fix for my colleague Walter tomorrow to have him run the original tests that were showing the issue.

mdavidsaver commented:
I'm starting to worry about adverse performance impacts.  This change would likely result in `eventsRemaining==0` more often, leading to more message fragmentation.  This could be mitigated by not setting `TCP_NODELAY`, which imo. would be reasonable for targets other than RTEMS and vxWorks (with coarse timer granularity).

mdavidsaver committed:
dbEvent: correct eventsRemaining

Only pass eventsRemaining when no queued events have been canceled.
Also possible race bt accessing ev_que with locking.

mdavidsaver committed:
dbEvent: try to detect possible "stall"

event_read() should not return if the last callback
was delivered with eventsRemaining!=0

mdavidsaver commented:
I think I have found a way avoid most of the excess message fragmentation by not passing `eventsRemaining==1` if any event on a queue has been canceled.  This may result in more missed opportunities for RSRV to combine messages than is strictly necessary.  The issue is only if that last event has been canceled.  On the other hand, assuming that cancelling subscriptions is relatively rare, the performance impact should be minimal.

I have also added a second commit which tries to warn in all cases where a `eventsRemaining==0` update is lost.  Including as described in https://github.com/epics-base/epics-base/issues/324#issuecomment-1399611177 .  As written `WARNING dbEvent possible queue stall` isn't very informative, but I'm not sure what information can usefully be printed.

ralphlange commented:
This fix works in the simplified and the original tests. (Run ~100 times with no failure.)

anjohnson commented:
Core Group Feb 15: Approved, needs a Release note entry to describe what changed.

mdavidsaver commented:
I've seen this new warning `WARNING dbEvent possible queue stall` appear once during IOC shutdown.  Probably spurious.  Need to investigate how the dbEvent worker exits...

mdavidsaver commented:
Changed to limit new warning message to once per dbEvent queue (aka. once per client w/ RSRV).

mdavidsaver commented:
Applied as b6626e4f60697d577097e335ce79e1ecbce3fee6
